File: VM7WC/vmss7wc16c3p2.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    vector<int> houses[2005];
    int distance[2005];
    bool visited[2005];
    int n, m, a, b;
    cin >> n >> m >> a >> b;
    for(int i=0; i<m; i++){
        int j, k;
        cin >> j >> k;
        houses[j].push_back(k);
        houses[k].push_back(j);
    }
    queue<int> q;
    q.push(a);
    visited[a] = true;
    distance[a] = 0;
    while(!q.empty()){
        int cur = q.front();
        q.pop();
        for(int i:houses[cur])
            if(visited[i]==false){
                q.push(i);
                visited[i] = true;
                distance[i] = distance[cur]+1;
            }
    }
    cout << (visited[b]?"GO SHAHIR!\n" : "NO SHAHIR!\n");
}

File: VM7WC/vmss7wc16c3p2.java
import java.util.*;

public class vmss7wc16c3p2 {
    
    public static void main(String[] args) {
        ArrayList<Integer>[] adj = new ArrayList[2001];
        boolean[] visited = new boolean[2001];
        
        Scanner scan = new Scanner(System.in);
        int n = scan.nextInt();
        int m = scan.nextInt();
        int a = scan.nextInt();
        int b = scan.nextInt();

        for(int i = 0; i < 2001; i++){
            adj[i] = new ArrayList<>();
        }

        for(int i=0; i<m; i++){
            int u = scan.nextInt();
            int v = scan.nextInt();
            adj[u].add(v);
            adj[v].add(u);
        }

        Queue<Integer> q = new LinkedList<>();
        visited[a] = true;
        q.add(a);

        while(!q.isEmpty()){
            int cur = q.poll();
            for(int nxt : adj[cur]){
                if(!visited[nxt]){
                    visited[nxt] = true;
                    q.add(nxt);
                }
            }
        }
        
        if(visited[b]){
            System.out.println("GO SHAHIR!");
        } else {
            System.out.println("NO SHAHIR!");
        }
        
        scan.close();
    }
}